### ISSUE TITLE
BUG: Indexing with two empty boolean masks yields a `(1,0)` size sparse matrix

### ISSUE THREAD

**Issue by ilan-gold** [scipy/scipy#19919]
### Describe your issue.

I would expect something like `mat[idx, idx_other_axis]` where `idx.sum() == 0` and `idx_other_axis.sum() == 0` to yield a size `(0,0)` matrix but it does not.  Sorry if this is not a bug, but the behavior doesn't match `numpy`, which is a `(0,)` sized matrix.  As I said, I think the closest analogy here is `(0,0)` in `scipy` then.  I apologize if this has an explanation but I don't know what it is and couldn't find anything in the repo about it.

### Reproducing Code Example

```python
from scipy import sparse
import numpy as np

mat = sparse.random(1_000, 1_000, density=0.01, format="csc")
idx = np.zeros((1000,), dtype=bool)
mat[idx, idx].shape # (1, 0)
mat = np.random.randn(1_000, 1_000)
mat_np[idx, idx].shape # (0,)
```

### Error message

```shell
N/A
```

### SciPy/NumPy/Python version and system information

```shell
1.11.4 1.26.2 sys.version_info(major=3, minor=11, micro=6, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /opt/arm64-builds/include
    lib directory: /opt/arm64-builds/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= SANDYBRIDGE MAX_THREADS=3
    pc file directory: /opt/arm64-builds/lib/pkgconfig
    version: 0.3.21.dev
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /opt/arm64-builds/include
    lib directory: /opt/arm64-builds/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= SANDYBRIDGE MAX_THREADS=3
    pc file directory: /opt/arm64-builds/lib/pkgconfig
    version: 0.3.21.dev
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.11.0
Compilers:
  c:
    commands: cc
    linker: ld64
    name: clang
    version: 14.0.0
  c++:
    commands: c++
    linker: ld64
    name: clang
    version: 14.0.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 0.29.36
  fortran:
    commands: gfortran
    linker: ld64
    name: gcc
    version: 12.1.0
  pythran:
    include directory: ../../pip-build-env-xp5pbjk1/overlay/lib/python3.11/site-packages/pythran
    version: 0.14.0
Machine Information:
  build:
    cpu: aarch64
    endian: little
    family: aarch64
    system: darwin
  cross-compiled: true
  host:
    cpu: aarch64
    endian: little
    family: aarch64
    system: darwin
Python Information:
  path: /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cibw-run-gft7gyfb/cp311-macosx_arm64/build/venv/bin/python
  version: '3.11'
```

**Comment by dschult** [comment]
This is independent of #19951 (see discussion there)

This error is controlled by [_index.py line 91](https://github.com/scipy/scipy/blob/e8b043df1d586d9df9fed96add1a536859f1cb8c/scipy/sparse/_index.py#L91).
`            return self.__class__(np.atleast_2d(row).shape, dtype=self.dtype)` when `row` is a size 0 array.

When handling index validation, what do you do if the array index has size zero?
The code creates an empty sparse matrix of the same type being indexed. But does an empty matrix have shape (1,0) or (0,0) or (0,1)?  They are all empty arrays of size 0.

Scipy sparse matrix chooses (1,0) by using essentially `np.atleast_2d([])`. Thus shape is (1,0).

Numpy has the flexibility to switch to 1d, stay at 2d or, in general make it nd. But they collapse any dimension that has length 0. I'm not sure why they decide to stick with 1d instead of going to 0d in this case. A 1d empty array has shape (0,).

This is the heart of this issue.  It presents this way for `csr, csc, dok, lil` formats for both sparse matrix and array -- basically anything that can be indexed has this difference between scipy-sparse and numpy.  

I think the question is "what should be returned from a blank mask?". It looks like [the array_api for boolean indexing](https://data-apis.org/array-api/2022.12/API_specification/indexing.html#boolean-array-indexing) specifies boolean index must be a single array. Combining it with another index in another dimension is prohibited. But it also admits that this creates "data dependent output" so some libraries will not be able to handle boolean indexes this way. So, the array api does not allow the form of indexing used in this Issue.

It remains that numpy and scipy.sparse both support this form of indexing with scipy.sparse returning a size 0 matrix of shape (1,0), and numpy returning a size 0 array of shape (0,).

Possible fixes:
- leave it as is... a 2d-matrix when indexed should return 2d and (1,0) is a 2d shape that has size 0.
- make it return a matrix with shape (0,0):   `self.__class__((0,0), dtype=self.dtype)`

More options exist for the new sparse arrays. There we can match numpy (produce a 1d array of size 0).
`self.__class__((0,), dtype=self.dtype)` which works in my upcoming `indexing-1d` branch (not yet a PR).

**Comment by ilan-gold** [comment]
> I think the question is "what should be returned from a blank mask?". It looks like [the array_api for boolean indexing](https://data-apis.org/array-api/2022.12/API_specification/indexing.html#boolean-array-indexing) specifies boolean index must be a single array. Combining it with another index in another dimension is prohibited. But it also admits that this creates "data dependent output" so some libraries will not be able to handle boolean indexes this way. So, the array api does not allow the form of indexing used in this Issue.

So I would be in favor of the second option because of the following (which is array-api legal):
```python
mask = np.zeros(2, dtype=bool)
assert np.array([[1, 2], [3, 4]])[mask, :][:, mask].shape == (0, 0)
```

Could I make the PR for this if you approve?

**Comment by dschult** [comment]
Can you find where this option is described as array-api legal?

The Note in [the array_api for boolean indexing](https://data-apis.org/array-api/2022.12/API_specification/indexing.html#boolean-array-indexing) says:

> An array must support indexing where the sole index is an M-dimensional boolean array B with shape S1 = (s1, ..., sM) according to the following rules. Let A be an N-dimensional array with shape S2 = (s1, ..., sM, ..., sN).

> **Note:** The prohibition against combining boolean array indices with other single-axis indexing expressions includes the use of None. To expand dimensions of the returned array, use repeated invocation of [expand_dims()](https://data-apis.org/array-api/2022.12/API_specification/generated/array_api.expand_dims.html#array_api.expand_dims).

So it looks like masks are not allowed to be combined with other indices as in `A[mask, :]`.  Am I reading this incorrectly? Or do i have the wrong array-api spec?

**Comment by ilan-gold** [comment]
@dschult You're correct, but I was more using that as an example of why an all zeros shape is the correct way to go.
```python
mask = np.zeros((2,2), dtype=bool)
np.array([[1, 2], [3, 4]])[mask].shape # (0,)
```
which is different again since
```
sparse.random(2, 2, density=0.01, format="csc")[mask].shape # (1,0)
```
I understand this change is not going to happen here, though, since we don't have 1d arrays yet but I still think `(1,0)` is strange.  To that point:
```
mask = np.zeros(2, dtype=bool)
np.array([[1, 2], [3, 4]])[mask].shape # (0,2)
```
This to me, this makes sense even if I don't like the 2 lying around (I would collapse to an empty matrix like you mention).

**Comment by ivirshup** [comment]
I would note that this behaviour isn't specific to boolean indexing and also occurs when indexing with empty lists. E.g.

```python
sparse.csr_array(np.arange(50).reshape(5, 10))[[], []].shape
# (1, 0)
```

Also, this is the behaviour for `np.matrix`:

```python
np.matrix(np.arange(25).reshape(5, 5))[[], []]
# matrix([], shape=(1, 0), dtype=int64)
```

So this may be intentional, at least for `sparse.spmatrix` subclasses. Though you can have an `np.matrix` with `shape = (0, 0)`

```python
np.matrix(np.ones((0, 0)))
# matrix([], shape=(0, 0), dtype=float64)
```

**Comment by ilan-gold** [comment]
Strange especially because your second example with boolean masks gives a different shape.  I am not sure what to do here and would be fine with nothing since this will hopefully be resolved via 1d arrays anyway and `numpy` is less consistent than I thought.
